utils/generator-utils.stanza

```diff
@@ -109,65 +109,39 @@ public defn connect-gnd-pins ():
         if port-name == Ref("gnd") :
           net (self.gnd, p)
 
-
-; Replace this once nets introspection is back
-public defn propagate-rail-voltage () :
-  ; println("Propagating rail voltages...")
-  inside pcb-module :
-    for connected-group in populated-connected-groups(self) do :
-      val power-supply-voltages = to-tuple $
-        for pin in connected-group seq? :
-          val prop = property?(pin.power-supply-pin)
-          match(value?(prop)) :
-            (p:PowerSupplyPin) : One(supply-voltage(p) as Double)
-            (_) : None() 
-      if not empty?(power-supply-voltages) :
-        val max-v = maximum(power-supply-voltages)
-        for pin in connected-group do:
-          property(pin.rail-voltage) = max-v
-
 public defn calculate-operating-points ()  :
   defn get-other-pin (comp:JITXObject,  p:JITXObject) :
     ; Used on 2-pin components to get the single other pin
     for ps in pins(comp) find! :
       ps != p
 
-  defn gnd-type-error! (inst, gnd) :
-    throw $ Exception("The gnd property of module %_ must be of type Pin or Net, got: '%_'." % [ref(inst), gnd])
-
   defn resistor-operating-point (v: Double, r: Double) :
     OperatingPoint(min-max(0.0, v), min-max(0.0, v / r))
 
   defn capacitor-operating-point (v: Double) :
     OperatingPoint(min-max(0.0, v), min-max(0.0, 0.0))
 
-  defn is-resistor? (i: Instance) :
+  defn is-resistor? (i: JITXObject) :
     has-property?(i.resistance)
 
-  defn is-capacitor? (i: Instance) :
+  defn is-capacitor? (i: JITXObject) :
     has-property?(i.capacitance)
 
-  defn resistance! (i: Instance) :
+  defn resistance! (i: JITXObject) :
     property(i.resistance)
 
   inside pcb-module :
-    val gnd = value? $ property?(self.gnd)
-
-    match(gnd: Pin|Net) :
-      ; println("Calculating operating points...")
-      val gnd-pins = connected-group(gnd)
-      for gnd-pin in gnd-pins do :
-        val inst = containing-instance!(gnd-pin)
-        if is-resistor?(inst) or is-capacitor?(inst) :
-          val other-pin = get-other-pin(inst, gnd-pin)
-          val voltage = value? $ property?(other-pin.rail-voltage)
-          match(voltage: Double) :
-            if is-resistor?(inst) :
-              property(inst.operating-point) = resistor-operating-point(voltage, resistance!(inst))
-            if is-capacitor?(inst) :
-              property(inst.operating-point) = capacitor-operating-point(voltage)
-    else :
-      gnd-type-error!(self, gnd) when gnd is-not False
+    for i in component-instances(self) do :
+      if has-property?(i.resistance) or has-property?(i.capacitance) or has-property?(i.inductance):
+        if length(pins(i)) == 2 :
+          if has-property?(pins(i)[0].net-voltage) and has-property?(pins(i)[1].net-voltage) :
+            val p1 = property(pins(i)[0].net-voltage)
+            val p2 = property(pins(i)[1].net-voltage)
+            val voltage = abs(max((max-value(p1) - min-value(p2)), (max-value(p2) - min-value(p1))))
```

Comment:
It seems the maximum of the 2 will always be positive. Actually, should we have a constructor check that ensures `tol-` and `tol+` are always positive? It seems it is very necessary and when I saw `Toleranced(Double, Double, Double)` I thought it was the usual `(min, norm, max)` but it is not

Reply:
You can have negative voltages